File: android/app/src/main/kotlin/com/example/a/ButtonListener.kt
package com.example.a

import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.concurrent.thread

class ButtonListener(private val ip: String, private val port: Int = 40004) {
    private var serverSocket: ServerSocket? = null
    private var clientSocket: Socket? = null
    private var isListening = AtomicBoolean(false)
    private var listenerThread: Thread? = null
    
    // Callback per notificare quando il pulsante viene premuto
    var onButtonPressed: (() -> Unit)? = null
    var onButtonReleased: (() -> Unit)? = null
    
    fun startListening() {
        if (isListening.get()) {
            println("[DEBUG] Button listener already running")
            return
        }
        
        isListening.set(true)
        
        listenerThread = thread {
            try {
                println("[DEBUG] Starting button listener on port $port in SERVER mode")
                // Utilizziamo principalmente la modalità server
                listenForConnections()
            } catch (e: Exception) {
                println("[ERROR] Failed to start button listener: ${e.message}")
                e.printStackTrace()
            } finally {
                stopListening()
            }
        }
    }
    
    // Modalità server: ascolta le connessioni in arrivo
    private fun listenForConnections() {
        try {
            println("[DEBUG] Creating server socket on port $port")
            // Crea un server socket che accetta connessioni da qualsiasi indirizzo
            serverSocket = ServerSocket(port, 50, null)
            
            println("[DEBUG] Server socket created successfully, waiting for connections")
            
            while (isListening.get()) {
                try {
                    println("[DEBUG] Waiting for connection on port $port")
                    val socket = serverSocket?.accept()
                    if (socket != null) {
                        println("[DEBUG] Received connection from ${socket.inetAddress.hostAddress}")
                        
                        // Gestisci la connessione in un thread separato
                        thread {
                            try {
                                handleConnection(socket)
                            } catch (e: Exception) {
                                println("[ERROR] Error handling connection: ${e.message}")
                                e.printStackTrace()
                            } finally {
                                try {
                                    socket.close()
                                } catch (e: Exception) {
                                    // Ignora errori di chiusura
                                }
                            }
                        }
                    }
                } catch (e: Exception) {
                    if (isListening.get()) {
                        println("[ERROR] Error accepting connection: ${e.message}")
                        e.printStackTrace()
                        // Breve pausa prima di riprovare
                        Thread.sleep(1000)
                    }
                }
            }
        } catch (e: Exception) {
            println("[ERROR] Failed to create server socket: ${e.message}")
            e.printStackTrace()
        }
    }
    
    private fun handleConnection(socket: Socket) {
        println("[DEBUG] Handling connection from ${socket.inetAddress.hostAddress}")
        
        try {
            val inputStream = socket.getInputStream()
            val buffer = ByteArray(10)
            
            // Continua a leggere dati finché la connessione è attiva
            while (isListening.get() && !socket.isClosed) {
                val bytesRead = inputStream.read(buffer)
                if (bytesRead <= 0) {
                    println("[DEBUG] End of stream reached")
                    break
                }
                
                println("[DEBUG] Received $bytesRead bytes")
                
                // Stampa i byte ricevuti per debug
                val hexString = buffer.take(bytesRead).joinToString(" ") { 
                    String.format("%02X", it) 
                }
                println("[DEBUG] Received data: $hexString")
                
                // Analizza il pacchetto - l'ottavo byte (indice 7) indica lo stato del pulsante
                if (bytesRead >= 8) {
                    val buttonState = buffer[7].toInt() and 0xFF
                    println("[DEBUG] Received button state: $buttonState")
                    
                    if (buttonState > 0) {
                        println("[DEBUG] Button pressed!")
                        onButtonPressed?.invoke()
                    } else {
                        println("[DEBUG] Button released!")
                        onButtonReleased?.invoke()
                    }
                }
            }
        } catch (e: Exception) {
            if (isListening.get()) {
                println("[ERROR] Error reading from socket: ${e.message}")
                e.printStackTrace()
            }
        }
    }
    
    fun stopListening() {
        isListening.set(false)
        try {
            println("[DEBUG] Stopping button listener")
            
            serverSocket?.close()
            serverSocket = null
            
            clientSocket?.close()
            clientSocket = null
            
            listenerThread?.join(1000)
            listenerThread = null
            
            println("[DEBUG] Button listener stopped")
        } catch (e: Exception) {
            println("[ERROR] Error stopping button listener: ${e.message}")
            e.printStackTrace()
        }
    }
} 

File: ios/Runner/ButtonListener.swift
import Foundation
import Network

class ButtonListener {
    private let ip: String
    private let port: Int
    private var isListening: Bool = false
    private var listenerThread: Thread?
    private var socket: Socket?
    private var serverSocket: NWListener?
    
    // Callback per notificare quando il pulsante viene premuto
    var onButtonPressed: (() -> Void)?
    var onButtonReleased: (() -> Void)?
    
    init(ip: String, port: Int = 40004) {
        self.ip = ip
        self.port = port
    }
    
    func startListening() {
        if isListening {
            NSLog("[DEBUG] Button listener already running")
            return
        }
        
        isListening = true
        
        listenerThread = Thread {
            NSLog("[DEBUG] Starting button listener on port \(self.port) in SERVER mode")
            self.listenForConnections()
        }
        
        listenerThread?.start()
    }
    
    private func listenForConnections() {
        let tcpOptions = NWProtocolTCP.Options()
        
        let parameters = NWParameters(tls: nil, tcp: tcpOptions)
        parameters.allowLocalEndpointReuse = true
        
        do {
            serverSocket = try NWListener(using: parameters, on: NWEndpoint.Port(integerLiteral: UInt16(port)))
            
            serverSocket?.stateUpdateHandler = { state in
                switch state {
                case .ready:
                    NSLog("[DEBUG] Server socket ready and listening on port \(self.port)")
                case .failed(let error):
                    NSLog("[ERROR] Server socket failed: \(error.localizedDescription)")
                    self.stopListening()
                default:
                    break
                }
            }
            
            serverSocket?.newConnectionHandler = { connection in
                NSLog("[DEBUG] Received connection from \(connection.endpoint)")
                
                // Gestisci la connessione
                connection.stateUpdateHandler = { state in
                    switch state {
                    case .ready:
                        NSLog("[DEBUG] Connection ready")
                        self.handleConnection(connection)
                    case .failed(let error):
                        NSLog("[ERROR] Connection failed: \(error.localizedDescription)")
                        connection.cancel()
                    case .cancelled:
                        NSLog("[DEBUG] Connection cancelled")
                    default:
                        break
                    }
                }
                
                connection.start(queue: .global())
            }
            
            serverSocket?.start(queue: .global())
            
            // Mantieni il thread in esecuzione finché isListening è true
            while isListening {
                Thread.sleep(forTimeInterval: 0.5)
            }
            
        } catch {
            NSLog("[ERROR] Failed to create server socket: \(error.localizedDescription)")
        }
    }
    
    private func handleConnection(_ connection: NWConnection) {
        func receiveNextMessage() {
            connection.receive(minimumIncompleteLength: 1, maximumLength: 10) { content, contentContext, isComplete, error in
                if let error = error {
                    NSLog("[ERROR] Error receiving data: \(error.localizedDescription)")
                    return
                }
                
                if let data = content, !data.isEmpty {
                    NSLog("[DEBUG] Received \(data.count) bytes")
                    
                    // Stampa i byte ricevuti per debug
                    let hexString = data.map { String(format: "%02X", $0) }.joined(separator: " ")
                    NSLog("[DEBUG] Received data: \(hexString)")
                    
                    // Analizza il pacchetto - l'ottavo byte (indice 7) indica lo stato del pulsante
                    if data.count >= 8 {
                        let buttonState = Int(data[7])
                        NSLog("[DEBUG] Received button state: \(buttonState)")
                        
                        if buttonState > 0 {
                            NSLog("[DEBUG] Button pressed!")
                            DispatchQueue.main.async {
                                self.onButtonPressed?()
                            }
                        } else {
                            NSLog("[DEBUG] Button released!")
                            DispatchQueue.main.async {
                                self.onButtonReleased?()
                            }
                        }
                    }
                }
                
                if isComplete {
                    NSLog("[DEBUG] Connection completed")
                    return
                }
                
                // Continua a ricevere dati
                if self.isListening {
                    receiveNextMessage()
                }
            }
        }
        
        receiveNextMessage()
    }
    
    func stopListening() {
        isListening = false
        
        serverSocket?.cancel()
        serverSocket = nil
        
        socket?.close()
        socket = nil
        
        listenerThread?.cancel()
        listenerThread = nil
        
        NSLog("[DEBUG] Button listener stopped")
    }
}

// Classe di supporto Socket per la compatibilità
class Socket {
    private var connection: NWConnection?
    
    init(host: String, port: Int) throws {
        let endpoint = NWEndpoint.hostPort(host: NWEndpoint.Host(host), port: NWEndpoint.Port(integerLiteral: UInt16(port)))
        connection = NWConnection(to: endpoint, using: .tcp)
        connection?.start(queue: .global())
    }
    
    func read(_ buffer: UnsafeMutablePointer<UInt8>, maxLength: Int) -> Int {
        var result = 0
        
        let semaphore = DispatchSemaphore(value: 0)
        
        connection?.receive(minimumIncompleteLength: 1, maximumLength: maxLength) { data, _, _, error in
            if let data = data, !data.isEmpty {
                data.copyBytes(to: buffer, count: min(data.count, maxLength))
                result = data.count
            } else if let error = error {
                NSLog("[ERROR] Socket read error: \(error.localizedDescription)")
                result = -1
            } else {
                result = 0
            }
            
            semaphore.signal()
        }
        
        _ = semaphore.wait(timeout: .now() + 2)
        return result
    }
    
    func write(_ data: Data) -> Int {
        var result = 0
        
        let semaphore = DispatchSemaphore(value: 0)
        
        connection?.send(content: data, completion: .contentProcessed { error in
            if let error = error {
                NSLog("[ERROR] Socket write error: \(error.localizedDescription)")
                result = -1
            } else {
                result = data.count
            }
            
            semaphore.signal()
        })
        
        _ = semaphore.wait(timeout: .now() + 2)
        return result
    }
    
    func close() {
        connection?.cancel()
        connection = nil
    }
} 
